README.md
# Alibaba Cloud C++ Software Development Kit
[中文文档](./README_zh.md)

[![GitHub version](https://badge.fury.io/gh/https%3A%2F%2Fgithub.com%2Faliyun%2Faliyun-openapi-cpp-sdk.svg)](https://badge.fury.io/gh/https%3A%2F%2Fgithub.com%2Faliyun%2Faliyun-openapi-cpp-sdk)
[![Build Status](https://travis-ci.org/aliyun/aliyun-openapi-cpp-sdk.svg?branch=master)](https://travis-ci.org/aliyun/aliyun-openapi-cpp-sdk)
[![codecov](https://codecov.io/gh/aliyun/aliyun-openapi-cpp-sdk/branch/master/graph/badge.svg)](https://codecov.io/gh/aliyun/aliyun-openapi-cpp-sdk)

The Alibaba Cloud C++ Software Development Kit (SDK) allows you to access Alibaba Cloud services such as Elastic Compute Service (ECS), Server Load Balancer (SLB), and CloudMonitor. You can access Alibaba Cloud services without the need to handle API related tasks, such as signing and constructing your requests.

This document introduces how to obtain and call Alibaba Cloud C++ SDK.

If you have any problem while using C++ SDK, please submit an [issue](https://github.com/aliyun/aliyun-openapi-cpp-sdk/issues/new).

## Prerequisites

- To use Alibaba Cloud C++ SDK, you must have an Alibaba Cloud account and an AccessKey.

	The AccessKey is required when initializing the client. You can create an AccessKey in the Alibaba Cloud console. For more information, see [Create an AccessKey](https://usercenter.console.aliyun.com/?spm=5176.doc52740.2.3.QKZk8w#/manage/ak).

	>**Note:** To increase the security of your account, we recommend that you use the AccessKey of the RAM user to access Alibaba Cloud services.

- To use Alibaba Cloud C++ SDK to access the APIs of a product, you must first activate the product on the [Alibaba Cloud console](https://home.console.aliyun.com/?spm=5176.doc52740.2.4.QKZk8w) if required.

- C++11 supported compiler installed
  - Windows: Visual Studio 2015 or newer
  - Linux: GCC 4.9 or newer
- CMake 3.0 or newer
- 4G memory or more

## Install C++ SDK from source code


### Linux

1. Install third-party libraries on the Linux platform, including `libcurl`, `libopenssl`, `libuuid`, and `libjsoncpp`.

	- Run the following commands on the `Redhat/Fedora` system
	```
    sudo dnf install libcurl-devel openssl-devel libuuid-devel libjsoncpp-devel
    ```
	- Run the following commands on the `Debian/Ubuntu` system
    ```
    sudo apt-get install libcurl4-openssl-dev libssl-dev uuid-dev libjsoncpp-dev
    ```
2. Run the following commands to clone source codes from GitHub.

    ```bash
    git clone https://github.com/aliyun/aliyun-openapi-cpp-sdk.git
    ```

3. Build and install SDK

  - Manually build and install

  ```bash
    cd aliyun-openapi-cpp-sdk
    mkdir sdk_build
    cd sdk_build
    cmake -DCMAKE_BUILD_TYPE=Release -DCMAKE_INSTALL_PREFIX=/usr ..
    make
    sudo make install
  ```

  - Or you may do this with `easyinstall.sh` in the source directory

  ```shell
    cd aliyun-openapi-cpp-sdk
    sudo sh easyinstall.sh
   ```

**The C++ SDK will be installed to `/usr`.**

### windows

1. Install [perl](https://www.perl.org/get.html#win32).

2. Run the following command to clone code from Github via git-bash:

  ```bash
  git clone https://github.com/aliyun/aliyun-openapi-cpp-sdk.git
  ```

2. Build Visual Studio solution

- Change directory to source code and make directory sdk_build

- Open CMake UI and

  - `Browse Source` to open source code directory.

  - `Browse build`  to open the created sdk_build directory

  - `Configure`

  - `Generate`

3. Build and Install C++ SDK

- Open `aliyun-openapi-cpp-sdk\\sdk_build\\alibabacloud-sdk.sln` with Visual Studio

- Select  `Release`

- Check INSTALL option from Build -> Configuration Manager

- Build->Build Solutions to build.

**SDK will be installed to `C:\Program File (x86)\alibabacloud-sdk`**


## Use the C++ SDK

Before using C++ SDK, you must first configure the preprocessor to define `ALIBABACLOUD_SHARED` to achieve dynamic linking with Alibaba Cloud C++ SDK shared libraries. Then you must create a client instance, specify the region of cloud services and provide authentication parameters before sending API requests.

The following code shows how to call the [DescribeInstances](~~25506~~) API of ECS to query detailed information of all ECS instances in a specific region.

```cpp
#include <iostream>
#include <alibabacloud/core/AlibabaCloud.h>
#include <alibabacloud/ecs/EcsClient.h>

using namespace AlibabaCloud;
using namespace AlibabaCloud::Ecs;

int main(int argc, char** argv) {
  // Initialize the SDK
  AlibabaCloud::InitializeSdk();

  // Configure the ECS instance
  ClientConfiguration configuration("<your-region-id>");
  EcsClient client("<your-access-key-id>", "<your-access-key-secret>", configuration);

  // Create an API request and set parameters
  Model::DescribeInstancesRequest request;
  request.setPageSize(10);

  auto outcome = client.describeInstances(request);
  if (!outcome.isSuccess()) {
    // Handle exceptions
    std::cout << outcome.error().errorCode() << std::endl;
    AlibabaCloud::ShutdownSdk();
    return -1;
  }

  std::cout << "totalCount: " << outcome.result().getTotalCount() << std::endl;

  // Close the SDK
  AlibabaCloud::ShutdownSdk();
  return 0;
}
```

Copy the above to ecs_test.cc, then build with the following command.

```bash
~$ g++ -o ecstest ecs_test.cc --std=c++11 -lalibabacloud-sdk-core -l alibabacloud-sdk-ecs
~$ ./ecstest
# Result or error message will be shown here.
~$
```

## Timeout Configuration

CPP SDK uses libcurl to do HTTP transfer.

- The following timeout parameters are used to for libcurl.

 - `connectTimeout`: timeout for the connect phase. [Refer](https://curl.haxx.se/libcurl/c/CURLOPT_CONNECTTIMEOUT_MS.html).
 - `readTimeout`: maximum time the request is allowed to take, [Refer](https://curl.haxx.se/libcurl/c/CURLOPT_TIMEOUT_MS.html)

- Default Value
 - `connectTimeout`: 5000ms
 - `readTimeout`: 10000ms

- You may specify `timeout` parameters when create a client or make a request.

- Request timeout has higher priority than client timeout.

- If you want to disable timeout feature, deliver `0` or `-1` to `setConnectTimeout` and `setReadTimeout`.

The following code shows hot to specify `timeout` parameters, and the final connectTimeout is 1000ms and readTimeout 6000ms.

```cpp
#include <iostream>
#include <alibabacloud/core/AlibabaCloud.h>
#include <alibabacloud/ecs/EcsClient.h>

using namespace AlibabaCloud;
using namespace AlibabaCloud::Ecs;

int main(int argc, char** argv) {
  // Initialize the SDK
  AlibabaCloud::InitializeSdk();

  // Configure the ECS instance
  ClientConfiguration configuration("<your-region-id>");
  // specify timeout when create client.
  configuration.setConnectTimeout(1500);
  configuration.setReadTimeout(4000);

  EcsClient client("<your-access-key-id>", "<your-access-key-secret>", configuration);

  // Create an API request and set parameters
  Model::DescribeInstancesRequest request;
  request.setPageSize(10);
  // specify timeout when request
  request.setConnectTimeout(1000);
  request.setReadTimeout(6000);

  auto outcome = client.describeInstances(request);
  if (!outcome.isSuccess()) {
    // Handle exceptions
    std::cout << outcome.error().errorCode() << std::endl;
    AlibabaCloud::ShutdownSdk();
    return -1;
  }

  std::cout << "totalCount: " << outcome.result().getTotalCount() << std::endl;

  // Close the SDK
  AlibabaCloud::ShutdownSdk();
  return 0;
}

```



**More [examples](https://github.com/aliyun/aliyun-openapi-cpp-sdk/tree/master/examples)**

## LICENSE
Please refer to [LICENSE](https://github.com/aliyun/aliyun-openapi-cpp-sdk/blob/master/LICENSE) (Apache 2.0 LICENSE).


CHANGELOG
2021-03-19 Version: 1.36.126
- Generated 2015-11-01 for `Ecsops`.

2021-03-14 Version: 1.36.125
- Generated 2019-02-25 for `sdk-test`.

2021-02-26 Version: patch
- Generated 2018-07-13 for `Ft`.

2021-02-26 Version: patch
- Generated 2018-07-13 for `Ft`.

2021-02-26 Version: patch
- Generated 2019-02-25 for `sdk-test`.

2021-02-26 Version: patch
- Generated 2019-02-25 for `sdk-test`.

2021-02-26 Version: patch
- Generated 2018-07-13 for `Ft`.

2021-02-26 Version: patch
- Generated 2018-07-13 for `Ft`.

2021-02-26 Version: patch
- Generated 2018-07-13 for `Ft`.

2021-02-26 Version: patch
- Generated 2018-07-13 for `Ft`.

2021-01-28 Version: patch
- Generated 2019-02-25 for `sdk-test`.

2021-01-28 Version: patch
- Generated 2019-02-25 for `sdk-test`.

2021-01-28 Version: patch
- Generated 2019-02-25 for `sdk-test`.

2021-01-28 Version: patch
- Generated 2019-02-25 for `sdk-test`.

2020-12-30 Version: patch
- Generated 2019-02-25 for `sdk-test`.

2020-12-30 Version: patch
- Generated 2019-02-25 for `sdk-test`.

2020-12-30 Version: patch
- Generated 2019-02-25 for `sdk-test`.

2020-12-30 Version: patch
- Generated 2019-02-25 for `sdk-test`.

2020-12-17 Version: patch
- Generated 2019-02-25 for `sdk-test`.

2020-12-17 Version: patch
- Generated 2019-02-25 for `sdk-test`.

2020-12-17 Version: patch
- Generated 2019-02-25 for `sdk-test`.

2020-12-17 Version: patch
- Generated 2019-02-25 for `sdk-test`.

2020-12-03 Version: patch
- Generated 2019-02-25 for `sdk-test`.

2020-12-03 Version: patch
- Generated 2019-02-25 for `sdk-test`.

2020-10-21 Version: patch
- Generated 2019-02-25 for `sdk-test`.

2020-10-21 Version: patch
- Generated 2019-02-25 for `sdk-test`.

2020-09-03 Version: patch
- Generated 2019-02-25 for `sdk-test`.

2020-08-26 Version: patch
- Generated 2018-07-13 for `Ft`.

2020-08-24 Version: patch
- Generated 2019-02-25 for `sdk-test`.

2020-08-24 Version: patch
- Generated 2019-02-25 for `sdk-test`.

2020-08-24 Version: patch
- Generated 2019-02-25 for `sdk-test`.

2020-08-20 Version: patch
- Generated 2018-07-13 for `Ft`.

2020-08-20 Version: patch
- Generated 2018-07-13 for `Ft`.

2020-08-20 Version: patch
- Generated 2018-07-13 for `Ft`.

2020-08-20 Version: patch
- Generated 2018-07-13 for `Ft`.

2020-08-20 Version: patch
- Generated 2019-02-25 for `sdk-test`.

2020-08-20 Version: patch
- Generated 2019-02-25 for `sdk-test`.

2020-08-20 Version: patch
- Generated 2019-02-25 for `sdk-test`.

2020-08-20 Version: patch
- Generated 2019-02-25 for `sdk-test`.

2020-08-20 Version: patch
- Generated 2019-02-25 for `sdk-test`.

2020-08-20 Version: patch
- Generated 2018-07-13 for `Ft`.

2020-08-20 Version: patch
- Generated 2019-02-25 for `sdk-test`.

2020-08-20 Version: patch
- Generated 2018-07-13 for `Ft`.

2020-08-20 Version: patch
- Generated 2018-07-13 for `Ft`.

2020-08-20 Version: patch
- Generated 2018-07-13 for `Ft`.

2020-08-20 Version: patch
- Generated 2018-07-13 for `Ft`.

2020-08-20 Version: patch
- Generated 2018-07-13 for `Ft`.

2020-08-20 Version: patch
- Generated 2018-07-13 for `Ft`.

2020-08-20 Version: patch
- Generated 2018-07-13 for `Ft`.

2020-08-20 Version: patch
- Generated 2018-07-13 for `Ft`.

2020-08-20 Version: patch
- Generated 2018-07-13 for `Ft`.

2020-08-20 Version: patch
- Generated 2018-07-13 for `Ft`.

2020-08-20 Version: patch
- Generated 2018-07-13 for `Ft`.

2020-08-20 Version: patch
- Generated 2018-07-13 for `Ft`.

2020-08-20 Version: patch
- Generated 2018-07-13 for `Ft`.

2020-08-20 Version: patch
- Generated 2018-07-13 for `Ft`.

2020-08-20 Version: patch
- Generated 2018-07-13 for `Ft`.

2020-08-20 Version: patch
- Generated 2018-07-13 for `Ft`.

2020-08-20 Version: patch
- Generated 2018-07-13 for `Ft`.

2020-08-20 Version: patch
- Generated 2018-07-13 for `Ft`.

2020-08-20 Version: patch
- Generated 2019-02-25 for `sdk-test`.

2020-08-20 Version: patch
- Generated 2019-02-25 for `sdk-test`.

2020-08-19 Version: patch
- Generated 2019-02-25 for `sdk-test`.

2020-08-19 Version: patch
- Generated 2019-02-25 for `sdk-test`.

2020-08-19 Version: patch
- Generated 2019-02-25 for `sdk-test`.

2020-08-19 Version: patch
- Generated 2019-02-25 for `sdk-test`.

2020-08-19 Version: patch
- Generated 2019-02-25 for `sdk-test`.

2020-08-19 Version: patch
- Generated 2019-02-25 for `sdk-test`.

2020-08-19 Version: patch
- Generated 2019-02-25 for `sdk-test`.

2020-08-19 Version: 1.36.124
- Generated 2019-02-25 for `sdk-test`.

2020-08-19 Version: patch
- Generated 2019-02-25 for `sdk-test`.

2020-08-19 Version: patch
- Generated 2019-02-25 for `sdk-test`.

2020-08-19 Version: patch
- Generated 2019-02-25 for `sdk-test`.

2020-08-19 Version: patch
- Generated 2018-07-13 for `Ft`.

2020-08-19 Version: patch
- Generated 2018-07-13 for `Ft`.

2020-08-18 Version: patch
- Generated 2018-07-13 for `Ft`.

2020-08-17 Version: patch
- Generated 2018-01-01 for `SDK`.

2020-08-17 Version: patch
- Generated 2018-01-01 for `SDK`.

2020-08-17 Version: patch
- Generated 2018-01-01 for `SDK`.

2020-08-17 Version: patch
- Generated 2018-01-01 for `SDK`.

2020-08-17 Version: patch
- Generated 2018-01-01 for `SDK`.

2020-08-17 Version: patch
- Generated 2018-01-01 for `SDK`.

2020-08-17 Version: patch
- Generated 2018-01-01 for `SDK`.

2020-08-17 Version: patch
- Generated 2018-01-01 for `SDK`.

2020-08-17 Version: patch
- Generated 2018-01-01 for `SDK`.

2020-08-17 Version: patch
- Generated 2019-02-25 for `sdk-test`.

2020-08-17 Version: patch
- Generated 2019-02-25 for `sdk-test`.

2020-08-13 Version: patch
- Generated 2019-02-25 for `sdk-test`.

2020-08-13 Version: patch
- Generated 2019-02-25 for `sdk-test`.

2020-08-13 Version: patch
- Generated 2019-02-25 for `sdk-test`.

2020-08-13 Version: patch
- Generated 2019-02-25 for `sdk-test`.

2020-08-12 Version: patch
- Generated 2019-02-25 for `sdk-test`.

2020-07-29 Version: patch
- Generated 2019-02-25 for `sdk-test`.

2020-07-27 Version: patch
- Generated 2019-02-25 for `sdk-test`.

2020-07-26 Version: patch

java/publish.js
const config = require('./setting');
const fs = require('fs');
const path = require('path');
const cp = require('child_process');
const promisify = require('util').promisify;
const exists = promisify(fs.exists);
const exec = promisify(cp.exec);
const readFile = promisify(fs.readFile);
const writeFile = promisify(fs.writeFile);
const mkdir = promisify(fs.mkdir);
const moment = require('moment');
const worker = require('./java_worker');
const ossutil = require('./oss_util');

async function msleep(n) {
  return new Promise(resolve => {
    setTimeout(() => {
      resolve();
    }, n);
  });
};

exports.clone = async function (repo, sdk, url) {
  if (await exists(sdk)) {
    // await exec('git pull origin master', { cwd: sdk });
    return;
  }
  await mkdir(repo, { recursive: true });
  console.log('git clone');
  console.log(repo);
  console.log(`git clone ${url} --depth 1 --no-checkout ${sdk}`);
  // clone sdk repo
  await exec(`git clone ${url} --depth 1 --no-checkout ${sdk}`, { cwd: repo });
  console.log('--------');
  // set git config
  await Promise.all([
    exec('git config user.email "sdk-team@alibabacloud.com"', { cwd: sdk }),
    exec('git config user.name "Alibaba Cloud SDK"', { cwd: sdk })
  ]);
}

exports.preBuild = async function (task, sdkCoreDir, options) {
  const output = task.output;
  await exec(`cp -rf ${sdkCoreDir} ${output}`, options);
  const cwd = path.join(output, path.basename(sdkCoreDir));
  await exec('git fetch origin master && ' +
    'git reset --hard origin/master && ' +
    'git checkout master',
    Object.assign({ cwd }, options));
}

exports.publishSdkToGithub = async function (task, gitPath, options, isRelease = false) {
  try {
    await release(task, gitPath, isRelease, options);
    const result = await exec('git rev-parse HEAD',
      Object.assign({ cwd: gitPath }, options));
    const commitid = result.stdout.replace('\n', '');
    const addresses = [
      `https://github.com/${task.gitbub_username}/${task.language}/commits/master`,
      `https://github.com/${task.gitbub_username}/${task.language}/commit/${commitid}`
    ];
    console.log(address);
    // await Promise.all([
    //   await releaseTask.addSuccessPackageReleaseTask(task, addresses[0], task.language, 'github'),
    //   await releaseTask.addSuccessPackageReleaseTask(task, addresses[1], task.language, 'git_diff'),
    //   sdkList.updateReleaseLog({ commit_url: addresses[0] }, task.taskId, task.language),
    // ]);
  } catch (err) {
    // ctx.logger.error(err);
    console.log(err);
    const address = `/task/${task.redisKey}/github?type=history`;
    console.log(address);
    //await this.updatePublishError(task, 'github', address, err);
  }
}

exports.publishSdkToOss = async function (task, targetPath) {
  const targetName = path.basename(targetPath);
  // ctx.logger.info(`push oss [${targetName}]: ${targetPath}`);
  await ossutil.upload(targetPath, `${config['ossDir']}/${targetName}`);
  const address = `https://${config['oss_bucket']}.${config['oss_region']}.aliyuncs.com/${config['ossDir']}/${targetName}`;
  console.log(address);
  //await releaseTask.addSuccessPackageReleaseTask(task, address, this.language, 'oss');
}

exports.publishToOssForDownload = async function (task, codeGenPath, options) {
  const productId = task.productId.toLowerCase();
  const targetName = `u-${task.taskId}-${task.dirPrefix}-${productId}.zip`;
  await ossutil.uploadZipDir(codeGenPath, targetName, 'downloads/', options);
  const address = `https://${config['oss_bucket']}.${config['oss_region']}.aliyuncs.com/downloads/${targetName}`;
  console.log(address);
  //await releaseTask.addSuccessPackageReleaseTask(task, address, this.language);
}

async function release(task, gitPath, release = false, options) {
  const { ctx } = this;
  options = Object.assign({ cwd: gitPath }, options);
  const repo = path.join(task.gitbub_username, task.repository);
  // Make remote address to token style for auth, token should from KMS.
  const { stdout: remote } = await exec('git remote -v', options);
  if (!~remote.indexOf(task.token)) {

    if (~remote.indexOf('origin')) {
      await exec('git remote rm origin', options);
    }
    await exec('git config user.name sdk-team && git config user.email sdk-team@alibabacloud.com', options);
    await exec(`git remote add origin https://sdk-team:${task.token}@github.com/${repo}.git`, options);
  }
  let tag, log;
  let i = 5;
  while (i-- > 0) {
    try {
      //确保代码最新
      try {
        await exec('git stash', options);
        await exec('git fetch origin master && git rebase origin/master', options);
        await exec('git stash pop', options);
      } catch (e) {
        // this.ctx.logger.error(e);
      }
      tag = await getReleaseTag(task);
      log = await updateChangeLog(task, gitPath);
      // For commit message invalid
      let commit = log.split('\n')[0].toString().replace(/`/g, '').replace('- ', '');
      // Push
      await gitPushCode(commit, options, task);
      break;
    } catch (err) {
      ctx.logger.error(err);
      if (i === 0) {
        throw err;
      }
      if (err.stderr && ~err.stderr.indexOf('.git/index.lock')) {
        // 因为index.lock报错,暂停30s
        await msleep(30 * 1000);
        continue;
      }
      await worker.conflictDeal(task, options, err);
      // 处理完冲突等15s再试
      await msleep(15 * 1000);
    }
  }

  if (!release) { return; }

  try {
    // Delete local tag
    await exec(`git tag -d ${tag}`, options);
    // Delete remote tag for generate success every time
    await exec(`git push origin --delete ${tag}`, options);
    // Delete remote release
    await exec('curl -v -i -X DELETE -H "Content-Type:application/json" -H' +
      `"Authorization: token ${task.token}" https://api.github.com/repos/${repo}/releases/${tag}'`, options);
  } catch (e) {
    // this.ctx.logger.error(e);
  }

  // Release with tag
  i = 6;
  while (i-- > 0) {
    try {
      //确保能打上tag
      await ctx.curl(`https://api.github.com/repos/${repo}/releases`, {
        method: 'POST',
        headers: {
          'Authorization': `token ${task.token}`,
          'Content-Type': 'application/json'
        },
        data: {
          tag_name: tag,
          target_commitish: 'master',
          name: tag,
          body: log,
          draft: false,
          prerelease: false
        }
      });
      break;
    } catch (err) {
      if (i === 0) {
        throw err;
      }
      continue;
    }
  }
}

async function gitPushCode(commit, options) {
  await exec('git add .', options);
  await exec(`git commit -m "${commit}"`, options);
  await exec('git push origin master', options);
}

async function updateChangeLog(task, gitPath) {
  const changelog = path.join(gitPath,
    `${task.dirPrefix}-${task.productId.toLowerCase()}`, 'ChangeLog.txt');
  const log = task.changeLog || '- Generated ' + task.apiVersions.join(', ') + ' for `' + task.productId + '`.';
  const changelogFileContent = await exists(changelog) ? await readFile(changelog, 'utf8') : '';
  let content = `${moment().format('YYYY-MM-DD')} Version: ${task.version}\n`
    + log + '\n\n'
    + changelogFileContent;
  await writeFile(changelog, content, 'utf8');
  return log;
}

async function getReleaseTag(task) {
  return `${task.productId}-${task.version}`;
}